libraries/chain/include/eosio/chain/webassembly/preconditions.hpp

```diff
@@ -0,0 +1,159 @@
+                                       std::is_same_v<decltype(is_unvalidated_ptr(std::declval<T>())), std::true_type> ||
+                                       std::is_same_v<name, T> ||
+                                       std::is_arithmetic_v<T>;
+      };
+      template <typename T>
+      struct is_whitelisted_type {
+         static constexpr bool value = std::is_arithmetic_v<std::decay_t<T>> ||
+                                       std::is_same_v<std::decay_t<T>, name> ||
+                                       std::is_pointer_v<T>    ||
+                                       std::is_lvalue_reference_v<T> ||
+                                       std::is_same_v<std::decay_t<T>, float128_t> ||
+                                       eosio::vm::is_span_type_v<std::decay_t<T>>;
+      };
+   }
+
+   template <typename T>
+   inline static constexpr bool is_whitelisted_type_v = detail::is_whitelisted_type<T>::value;
+
+   template <typename T>
+   inline static constexpr bool is_whitelisted_legacy_type_v = detail::is_whitelisted_legacy_type<T>::value;
+
+   template <typename... Ts>
+   inline static constexpr bool are_whitelisted_types_v = true; //(... && detail::is_whitelisted_type<Ts>::value);
+
+   template <typename... Ts>
+   inline static constexpr bool are_whitelisted_legacy_types_v = (... && detail::is_whitelisted_legacy_type<Ts>::value);
+
+   template <typename T, typename U>
+   inline static bool is_aliasing(const T& a, const U& b) {
+      std::uintptr_t a_ui = reinterpret_cast<std::uintptr_t>(a.data());
+      std::uintptr_t b_ui = reinterpret_cast<std::uintptr_t>(b.data());
+      return a_ui < b_ui ? a_ui + a.size() >= b_ui : b_ui + b.size() >= a_ui;
+   }
+   inline static bool is_nan( const float32_t f ) {
+      return f32_is_nan( f );
+   }
+   inline static bool is_nan( const float64_t f ) {
+      return f64_is_nan( f );
+   }
+   inline static bool is_nan( const float128_t& f ) {
+      return f128_is_nan( f );
+   }
+
+   EOS_VM_PRECONDITION(early_validate_pointers,
+         EOS_VM_INVOKE_ON_ALL([&](auto&& arg, auto&&... rest) {
+            using namespace eosio::vm;
+            using arg_t = std::decay_t<decltype(arg)>;
+            if constexpr (is_reference_proxy_type_v<arg_t>) {
+               if constexpr (is_span_type_v<dependent_type_t<arg_t>>) {
+                  using dep_t = dependent_type_t<arg_t>;
+                  const dep_t& s = arg;
+                  EOS_ASSERT( s.size() <= std::numeric_limits<wasm_size_t>::max() / (wasm_size_t)sizeof(dependent_type_t<dep_t>),
+                        wasm_execution_error, "length will overflow" );
+                  volatile auto check = *(reinterpret_cast<const char*>(arg.original_ptr) + s.size_bytes() - 1);
+                  ignore_unused_variable_warning(check);
+               } else {
+                  EOS_ASSERT(arg.original_ptr != ctx.get_interface().get_memory(), wasm_execution_error, "references cannot be created for null pointers");
```

Comment:
Even though there was a specialization to handle pointer arguments in the old intrinsic system, no intrinsics actually used it.  All legacy_ptr's correspond to what were reference arguments, for which this behavior is correct.

Reply:
So the conclusion we reached is that the `early_validate_pointers` precondition can be replaced by a `null_pointer_check` precondition that validates for any `legacy_ptr` that it does point to the start of the WASM linear memory (to maintain same behavior will all the reference arguments in the old intrinsic system).

Furthermore, we don't care to impose such an unnecessary null check on new host functions going forward that want to be able to modify the values (stored in WASM linear memory) passed to them. Rather than using references to handle such cases (for example out variables), we want to enforce new host functions to be explicit about it by using a pointer directly (even though the pointer will always be pointing to a valid region in WASM linear memory). And for these new pointers we don't want to impose validation like in `null_pointer_check`.

This does mean that some of the functions mentioned below which are registered as non-legacy but still use a reference need to be converted into using a `legacy_ptr` instead and be registered using the legacy version of the macros.